### PR TITLE
Prefix `target_exe` with `setup` dir at use sites

### PR DESCRIPTION
**Pull request by ranweiler**
Stop server-side prefixing of task `target_exe` parameters with a hardcoded `setup/` relative path. Instead, prefix `target_exe` in each task implementation, using the absolute path of the local mirror of the `setup` container.

Note that this preserves the existing end-user API and internal data model, but internally enforces the following invariant: the `target_exe` field that occurs in `TaskDetails` and `TaskUnitConfig` now match, and are always a `setup`-relative virtual path (blob name). Users of `target_exe` (like the `onefuzz-task` runner) must combine this relativized subpath with a path to a local `setup` directory to concretely locate the `target_exe`.

This change will effect the following breakage on upgrade: old `TaskUnitConfig`'s sitting in a work queue as part of a work set will have a `target_exe` that has been pre-prefixed with `setup/`. If an upgraded agent picks up the old task, it will result in double-prefixing the `target_exe` at runtime, causing the task to fail.

See larger long-term design discussion in #2399.

### CODE REVIEW

File: src/ApiService/ApiService/onefuzzlib/Config.cs [hunk 1]
```diff
@@ -257,9 +257,8 @@ await _containers.GetContainerSasUrl(x.Item2.Name, StorageType.Corpus, ConvertPe
         }
 
         if (definition.Features.Contains(TaskFeature.CoverageFilter)) {
-            var coverageFilter = task.Config.Task.CoverageFilter;
             if (coverageFilter != null) {
-                config.CoverageFilter = $"setup/{coverageFilter}";
+                config.CoverageFilter = task.Config.Task.CoverageFilter;
```
Comment (ranweiler) [inline src/ApiService/ApiService/onefuzzlib/Config.cs:261]: This change has not been accounted for the VM agent, addressing now.

File: src/agent/onefuzz-task/src/tasks/analysis/generic.rs [hunk 2]
```diff
@@ -194,11 +194,13 @@ pub async fn run_tool(
     config: &Config,
     reports_dir: &Option<PathBuf>,
 ) -> Result<()> {
+    let target_exe = config.common.setup_dir.join(&config.target_exe);
```
Comment (ranweiler) [inline src/agent/onefuzz-task/src/tasks/analysis/generic.rs:197]: To avoid the upgrade breakage described in the PR message, we could temporarily guard this `.join()` on a call to `.starts_with("setup/")`. This does have some edge cases, though, e.g. if a user had a `setup` container which had a nested `target_exe` that (abstractly) resided in `{setup_dir}/setup/fuzz.exe`.

File: src/cli/onefuzz/templates/libfuzzer.py [hunk 3]
```diff
@@ -550,7 +550,7 @@ def dotnet(
         if target_options is None:
             target_options = []
         target_options = [
-            "--target_path={setup_dir}/" f"{target_harness}"
```
Comment (ranweiler) [inline src/cli/onefuzz/templates/libfuzzer.py:553]: Opportunistic fix from other PR, which I closed without merging.

File: src/agent/onefuzz-task/src/tasks/coverage/generic.rs [hunk 4]
```diff
@@ -138,9 +138,34 @@ impl CoverageTask {
...
 
+        let setup_dir = &self.config.common.setup_dir;
+
         // Ensure users can locate the filter relative to the setup container.
-        let expand = Expand::new().setup_dir(&self.config.common.setup_dir);
-        let filter_path = expand.evaluate_value(raw_filter_path)?;
+        let expand = Expand::new().setup_dir(setup_dir);
+        let expanded_filter_path = expand.evaluate_value(raw_filter_path)?;
+
+        // Determine if `coverage_filter` is already prefixed with a absolute or relative
```
Comment (ranweiler) [inline src/agent/onefuzz-task/src/tasks/coverage/generic.rs:147]: Hmm, we can do this more robustly by introducing a `find_setup_relative_file()` function which could use file existence checks and attempted expansions to handle the cases of no prefixing (a fully relativized subpath), `setup/`-prefixing, and `{setup_dir}`-prefixing. We could use that same function wherever we are prepending `{setup_dir}` to `target_exe`. Let's try that instead.

File: src/agent/onefuzz-task/src/tasks/analysis/generic.rs (hunk 2 above)
Comment (tevoinea) [reply src/agent/onefuzz-task/src/tasks/analysis/generic.rs:197]: I think the guard makes sense -- we go from guaranteed breakage of all inflight 
 WorkUnits  to possible/unlikely breakage. The only thing is, after this change gets deployed, we should remove the guard in the following deploy so we don't keep this edge case around.

**Review by tevoinea — APPROVED** [review]

File: src/agent/onefuzz-task/src/tasks/coverage/generic.rs [hunk 5]
```diff
@@ -138,9 +139,17 @@ impl CoverageTask {
...
+        let resolved =
+            resolve_setup_relative_path(&self.config.common.setup_dir, raw_filter_path).await?;
+        let filter_path = if let Some(path) = resolved {
+            path
+        } else {
+            error!(
+                "unable to resolve setup-relative coverage filter path: {}",
+                raw_filter_path
+            );
+            return Ok(CmdFilter::default());
```
Comment (ranweiler) [inline src/agent/onefuzz-task/src/tasks/coverage/generic.rs:151]: Note the behavior here: if a `coverage_filter` was specified, but we couldn't find it, fall back to the default (empty) filter.

Alternatively, we may just fail the task. Currently, this is the only way to proactively inform the user that their `coverage_filter` subpath was invalid (present, but not resolvable under `setup_dir`).

File: src/agent/onefuzz-task/src/tasks/fuzz/libfuzzer/common.rs [hunk 6]
```diff
@@ -236,12 +236,7 @@ where
                 .for_each(|d| inputs.push(&d.local_path));
         }
 
-        let fuzzer = LibFuzzer::new(
-            &self.config.target_exe,
-            self.config.target_options.clone(),
-            self.config.target_env.clone(),
-            &self.config.common.setup_dir,
-        );
+        let fuzzer = L::from_config(&self.config).await?;
```
Comment (ranweiler) [inline src/agent/onefuzz-task/src/tasks/fuzz/libfuzzer/common.rs:239]: Opportunistic fix. We should only ever be creating `LibFuzzer` values via `L::from_config()`.

File: src/agent/onefuzz-task/src/tasks/utils.rs [hunk 7]
```diff
@@ -96,3 +96,157 @@ pub fn parse_key_value(value: String) -> Result<(String, String)> {
...
+            anyhow::format_err!(
+                "unable to resolve subpath `{}` under setup dir `{}`",
+                subpath.display(),
+                setup_dir.display()
+            )
+        })
+}
+
+/// Try to resolve an ambiguous setup-relative subpath, returning `None` if not found.
+pub async fn resolve_setup_relative_path(
```
Comment (ranweiler) [inline src/agent/onefuzz-task/src/tasks/utils.rs:120]: @tevoinea, let me know if you have any questions about this behavior.

One thing I just noticed: I think a (valid, existing) absolute path will result in `Ok(None)`. We should probably check for that.

Comment (ranweiler) [reply src/agent/onefuzz-task/src/tasks/utils.rs:120]: (that is: we should either reject any absolute `subpath`, or alternatively, if `subpath` is absolute, and `setup_dir` is already a prefix of `subpath`, then this should arguably succeed)

File: src/agent/onefuzz-task/src/tasks/utils.rs [hunk 8]
```diff
@@ -96,3 +96,157 @@ pub fn parse_key_value(value: String) -> Result<(String, String)> {
...
+
+        test_case(RELATIVE, "x/y/fuzz.exe").await?;
+        test_case(RELATIVE, "setup/x/y/fuzz.exe").await?;
+        test_case(RELATIVE, "{setup_dir}/x/y/fuzz.exe").await?;
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn test_resolve_setup_relative_path_nested_setup() -> Result<()> {
```
Comment (ranweiler) [inline src/agent/onefuzz-task/src/tasks/utils.rs:243]: This tests the unlikely edge case that we have something like a _new_, non-legacy `target_exe` that is prefixed by `setup/` because the target genuinely exists in a path like `{setup_dir}/setup/fuzz.exe`.

File: src/agent/onefuzz-task/src/tasks/utils.rs (hunk 7 above)
Comment (tevoinea) [reply src/agent/onefuzz-task/src/tasks/utils.rs:120]: What do you think of logging a warning if more than one of these resolve? I imagine it's _probably_ a misconfiguration and the user would want to know.

Comment (ranweiler) [reply src/agent/onefuzz-task/src/tasks/utils.rs:120]: I'm going to stick with "return the first resolved" for now, because I don't actually want to actually attempt every single case. The main goal here is to have have a comprehensible resolution rule/order to aid debugging.

**Review by tevoinea — APPROVED** [review]
